### ISSUE TITLE
HTTPS by default on openstreetmap.org

### ISSUE THREAD

**Issue by grischard** [openstreetmap/operations#117]
This ticket is intended to be a central place to discuss the possibility of switching openstreetmap.org to be https by default.

*This is a subject many people feel passionate about*, one way or the other, and not a switch that can simply be flicked on and off. Everyone's concerns, issues on the way, hopes, desires, possible solutions, etc. should be discussed or linked here.

Wikipedia did the switch [in 2015](https://blog.wikimedia.org/2015/06/12/securing-wikimedia-sites-with-https/), The Guardian [in 2016](https://www.theguardian.com/info/developer-blog/2016/nov/29/the-guardian-has-moved-to-https).

This isn't about making all content available over https - that's AFAIK already the case, and users of the HTTPS Everywhere will always use HTTPS when loading the website.

The questions that we should address are:

## What advantages would a switch to HTTPS by default bring?

A few so far:

- [Un-breaking "Show my location"](https://github.com/openstreetmap/operations/issues/117#issuecomment-258642954) on [many browsers](https://github.com/openstreetmap/operations/issues/117#issuecomment-258836089) without [needing any development](https://github.com/openstreetmap/operations/issues/117#issuecomment-258798235).
- Easing the load on the servers if we can serve all content over HTTP/2 (but that's [already enabled for the tile servers, which represent most of the global load](https://github.com/openstreetmap/operations/issues/117#issuecomment-258642954))
- Better security and privacy for everyone - see the reasons that drove Wikipedia to switch. Make logged in users [safer from cookie-stealing attacks](https://github.com/openstreetmap/operations/issues/117#issuecomment-258798235).
- More reliable Referer headers - https pages don't send them to http pages by default

## What issues would a switch to HTTPS by default cause?

* People on bad connections might have trouble loading the site.
  * @systemed [is concerned that HTTPS connections are less reliable over bad mobile links](https://github.com/openstreetmap/operations/issues/117#issuecomment-258699479).
  * **Can anyone show cases where openstreetmap.org is reproducibly slower or less reliable over https**?
  * Is this something that can be mitigated or improved on the OWG side, for example by enabling http/2 everywhere?
  * I couldn't reproduce the issue by loading https://www.httpvshttps.com on a simulated lossy Edge network. (5% packet loss). The tile servers upgrade to http/2 on modern browsers anyway. Unless anyone can show that this effect exists, it sounds a lot like superstition.

* People using OSM where HTTPS is blocked couldn't navigate the site.
  * petschge (who's on IRC but not on github) was at a private university in Germany in 2015 where a faulty proxy made https browsing impossible. Unfortunately, he's not in contact with them anymore, and can't confirm that the problem still exists. Similarly, @Zverik [was in a classroom where HTTPS was blocked](https://github.com/openstreetmap/openstreetmap-website/pull/939#issuecomment-84707915) in the past. A lot has happened since then: many major sites, e.g. wikipedia and github, now use HSTS, and Wikipedia's 2015 switch might have forced educational institutions to fix this.
  * **Does anyone know of users who can't access openstreetmap.org over https at all?**
  * A possible way of measuring this would be by including two identical tracker pixels on openstreetmap.org, one served over http and the other over https, and compare the hit counts after a week.
  * Logging in requires https in any case.

* Legacy browsers would be locked out.
  * [SSLLabs says that this would break IE6 on XP and Java 6 users](https://www.ssllabs.com/ssltest/analyze.html?d=openstreetmap.org&latest), but that was [a tiny proportion of users](https://github.com/openstreetmap/operations/issues/117#issuecomment-258719144) in November 2016.
  * This could also be something that a tracker pixel and a detailed analysis of the user agents in the logs could provide

* [Content that the iD editor embeds from other sites must be served over https](https://github.com/hotosm/osm-tasking-manager2/issues/988). But that's already the case, since not doing so breaks the embed for users who are already editing over https.

## Should we redirect any traffic to HTTPS, or only strongly encourage it?

[Forcing and strongly encouraging are different things](https://github.com/openstreetmap/operations/issues/117#issuecomment-258719144).

Sending HSTS headers, converting all links to https://, or both, would move _most_ traffic to HTTPS while still replying to legacy HTTP requests. Sending 301 redirects would make sure _every_ client is using https, but make it impossible to use http.

## What issues would HSTS cause?

[HSTS breaks any site that accesses the OAuth API with http URLs](https://github.com/openstreetmap/openstreetmap-website/pull/1341#issuecomment-258439266).
[@mojodna has offered to patch our OAuth library](https://github.com/openstreetmap/openstreetmap-website/issues/1493#issuecomment-327238826) to mitigate the issue.

HSTS could also be selectively enabled, for example only for the http/2 enabled tile servers, which could make the loading of maps faster. Note that clients that are http/2 capable automatically upgrade to https when they connect to a http/2 capable server.

## What would technically need to be done to make the switch?

Many pieces of our infrastructure would probably need to be changed or configured differently. How much work would this be for the OWG? [Andy is saying it's doable](https://github.com/openstreetmap/operations/issues/117#issuecomment-258676164).

## Considering the pluses and minuses, is it worth it?

Many of us, myself included, have (strong) opinions on this. I would like us to reach a decision based on measurable facts, and hypothetically review that decision if those facts change over time.

*This list of questions is not meant to be exhaustive*. I'll try to update the ticket to maintain visibility as comments come in.

**Comment by grischard** [comment]
CC @tyll @gravitystorm @bhousel @rorym @tomhughes @genodeftest because you've all shown interest in this before.

**Comment by tomhughes** [comment]
Ignore IE6 and Java 6 as they are already unsupported - you can't login using them to start with as that already requires https.

**Comment by pnorman** [comment]
> Easing the load on the servers if we can serve all content over HTTP/2?

Probably not significant. Tiles are over HTTP/2 when possible, and that's where the many is particularly useful.

> What issues would HSTS cause

I consider HSTS an implementation detail to the question of requiring all osm.org traffic to use HTTPS. How to implement is best discussed after the policy decision is made.

---
A couple of years ago I'd have disagreed, but HTTP/2 and overall HTTPS support have reduced the costs to requiring HTTPS, and Chrome has made a big change:

![image](https://cloud.githubusercontent.com/assets/1190866/20033752/d3d5b302-a365-11e6-9e05-346c56755142.png)

This means the "Show my Location" button doesn't work in Chrome over HTTP. To be clear, I think what Chrome is doing with blocking the location API is wrong, not addressing a real thread, and doesn't follow [their own proposal](https://www.chromium.org/Home/chromium-security/prefer-secure-origins-for-powerful-new-features) because it's not a new feature. But we're stuck with it. The fix for this is to redirect any page on openstreetmap.org that shows a map to HTTPS. This is most pages.

I think the advantage of getting geolocation working in Chrome outweighs any current disadvantages of requiring HTTPS.

**Comment by gravitystorm** [comment]
> This means the "Show my Location" button doesn't work in Chrome over HTTP

This is also the case for Safari on iOS, and is a bit annoying and I doubt many people can figure it out.

Given that you can't log into the site without HTTPS, any use-case that involves mis-configured proxies etc aren't a high priority. However, I'm interested to hear more about slow connections and how it impacts, and to see if this is still a problem.

> How much work would this be for the OWG?

It's not trivial, but I don't think it's excessively large. If we decide to make the policy for https-everywhere, then we can work through whatever needs doing at a steady rate.

**Comment by grischard** [comment]
> Ignore IE6 and Java 6 as they are already unsupported - you can't login using them to start with as that already requires https.

You're right, no one edits with them anymore, but are they also insignificant for the tile server user agent statistics?

**Comment by grischard** [comment]
https://wikitech.wikimedia.org/wiki/HTTPS and https://wikitech.wikimedia.org/wiki/HTTPS/Future_work and https://phabricator.wikimedia.org/T104681 and https://phabricator.wikimedia.org/tag/https/ might be interesting references.

**Comment by systemed** [comment]
> @systemed is concerned that HTTPS connections are less reliable over bad links

I'll opt out of this discussion, I'm afraid. I have consistently been in areas of poor mobile coverage where https:// sites wouldn't load but http:// would (in particular, the boatyard in Burton-on-Trent where I moored for several years).

However, the arguments used for https are sufficiently emotive, and the groupthink sufficiently strong (cf eight gazillion discussions on HN and elsewhere), that I have no great interest in wasting hours on this that could be more productively spent elsewhere. As long as there are other mapping sites still available over http:// which I can use in areas of poor reception, I no longer have the energy to argue as to what OSM might do. But thank you for the mention.

**Comment by zerebubuth** [comment]
> You're right, no one edits with them anymore, but are they also insignificant for the tile server user agent statistics?

IE6 (or user-agents claiming to be it) accounted for 0.16% of tile hits yesterday.

> @simonpoole also expressed concern about Android users on version 2.2 or less. I don't know if you can get that from the user agents.

As far as I can tell Android <= 2.2 accounts for about 0.01% of site visitors and 0.005% of tile views according to user-agent headers yesterday. It is likely that there are apps running on Android versions <= 2.2 which wouldn't show up in this analysis, as they don't report the OS version, but I'd be surprised if it were a large proportion.

On a different subject, the title of this ticket is "HTTPS by default", but much of the discussion has been around HSTS, "requiring all osm.org traffic to use HTTPS" or "converting all links to https://". To me, these are very different things. "By default" doesn't shut the door to people who make the choice to use HTTP, whereas HSTS or converting all links to https:// would.

As pointed out, it is _already_ the situation that logging in or other secure actions require HTTPS. It is _already_ the situation that anyone choosing to use HTTPS Everywhere uses the site in a fully secure manner.

The only real problem appears to be the fairly minor issue that "show my location" doesn't work for some recent browsers. Would it be possible to "un-break" Chrome and Safari by detecting the failure condition and reloading the page over HTTPS?

**Comment by grischard** [comment]
You're right @zerebubuth, these are different things, and identifying the possible tools we have is essential to decide which ones to use.

FWIW, http://caniuse.com/#search=hsts says that switching on HSTS would affect 90% of the users.

**Comment by grischard** [comment]
It would still be very interesting to know under what circumstances https://openstreetmap.org is broken while http:// isn't.

Something like https://github.com/jonocole/disable-hsts-chrome-extension could be useful to @systemed and other hypothetical affected users.

**Comment by pnorman** [comment]
> On a different subject, the title of this ticket is "HTTPS by default", but much of the discussion has been around HSTS, "requiring all osm.org traffic to use HTTPS" or "converting all links to https://". To me, these are very different things. "By default" doesn't shut the door to people who make the choice to use HTTP, whereas HSTS or converting all links to https:// would.

Using HTTPS by default to me means that when someone types in openstreetmap.org they end up on the HTTPS site. The only ways to accomplish that I know of are direct all traffic requesting the front page to HTTPS. That could be by means of HSTS, 301 redirects, or similar. This will effectively make it impossible or very difficult to browse openstreetmap.org with HTTP.

> The only real problem appears to be the fairly minor issue that "show my location" doesn't work for some recent browsers. Would it be possible to "un-break" Chrome and Safari by detecting the failure condition and reloading the page over HTTPS?

About 75% of browsers require HTTPS for geolocation, and I suspect we'd have to have them redirected to HTTPS on page load, or press the geolocate button twice.

**Comment by tyll** [comment]
On Sun, Nov 06, 2016 at 03:17:25PM -0800, Matt Amos wrote:

> On a different subject, the title of this ticket is "HTTPS by
> default", but much of the discussion has been around HSTS, "requiring
> all osm.org traffic to use HTTPS" or "converting all links to
> https://". To me, these are very different things. "By default"
> doesn't shut the door to people who make the choice to use HTTP,
> whereas HSTS or converting all links to https:// would.

HSTS does not mean that it is impossible to use the site with HTTP. It
only means that once someone accessed the site via HTTPS they will
continue to use HTTPS for a certain amount of time. You can even use
this to upgrade anyone to HTTPS if their browser supports HTTPS for
example by including a hidden image via HTTPS and serving a HSTS header
via HTTPS.

Also even if all links from the outside (e.g. in mails) are https,
people who prefer http can still use it. However, the future is that
everything will be https and that will not be able to use the big sites
on the Internet without https.

> As pointed out, it is _already_ the situation that logging in or other
> secure actions require HTTPS. It is _already_ the situation that
> anyone choosing to use HTTPS Everywhere uses the site in a fully
> secure manner.

IMHO it should be a project's resposibility to make sure that their
users can and usually access a site in a secure manner. This protects
everyone who does not have the expertise to know whether or not what
they are doing is secure.

**Comment by zerebubuth** [comment]
> About 75% of browsers require HTTPS for geolocation, and I suspect we'd have to have them redirected to HTTPS on page load, or press the geolocate button twice.

I was thinking that, if the failure can't be detected until the button is pushed, then we'd have to pop up some sort of `div` with a message saying "This feature is only available over HTTPS, please click this link to reload."

Alternatively, if it's possible to detect before the button is pressed, then the button could be greyed out with a tooltip explaining why it's greyed out, possibly also with a link to reload over HTTPS.

> Using HTTPS by default ... will effectively make it impossible or very difficult to browse openstreetmap.org with HTTP.

I think so too. While it would technically be possible to copy each link by hand into the address bar and edit the scheme to "http"... that would be so much work that no one would do it.

In summary so far:

**Pros:**

* No development needed to "un-break" Chrome & Safari.
* People using OSM over HTTP over public WiFi are safer from cookie-stealing attacks. (Although we could mitigate this with the [secure attribute on cookies](https://tools.ietf.org/html/rfc6265#section-4.1.2.5))

**Cons:**

* People on slow or mobile connections might find it difficult to load the site.
* People using OSM over networks where HTTPS is interdicted would find it unusably difficult to navigate the site (although they would not have been able to log in anyway).

**Comment by pnorman** [comment]
> we'd have to pop up some sort of div with a message

I'd rather redirect, I think it's a much better user experience. A div popup turns a one-click operation into a three-click operation.

> In summary so far: ...

What about the API calls? These have different usage patterns and user-agents, and they get into OAuth signing. Or are we only looking at the HTML pages right now?

**Comment by genodeftest** [comment]
> The only real problem appears to be the fairly minor issue that "show my location" doesn't work for some recent browsers. Would it be possible to "un-break" Chrome and Safari by detecting the failure condition and reloading the page over HTTPS?

http://caniuse.com/#search=geolocation says:
HTTPS is required for Chrome, Safari, Opera, iOS Safari, Android Browser, Chrome for Android. Only major browsers without this requirement are IE, Edge and Firefox.

**Comment by grinapo** [comment]
Just a sidenote: a site got burned by lack of entropy in the entropy pool when using it up for all kinds of crypto at a changeover. 
If one's using blocking random access (rejecting the use of "bad" quality randoms when there isn't enough entropy available) it simply blocks for extended periods of time. Non-blocking randoms with no entropy mean deteriorated security (without the blocking, obviously :-)).
Usually a non-issue for "normal" configs, but just in case keep an eye on the gauge.

**Comment by tyll** [comment]
On Mon, Nov 07, 2016 at 10:28:30AM -0800, grinapo wrote:

> If one's using blocking random access (rejecting the use of "bad"
> quality randoms when there isn't enough entropy available) it simply
> blocks for extended periods of time. Non-blocking randoms with no
> entropy mean deteriorated security (without the blocking, obviously
> :-)).

Do you have any proof that /dev/urandom (aka the non-blocking random
pool) is less secure than /dev/random? This talk explains why
/dev/urandom is the right random source:
https://www.youtube.com/watch?v=Q8JAlZ-HJQI&t=1337

**Comment by Firefishy** [comment]
We enabled haveged across all services a few years ago to avoid entropy pool starvation.

**Comment by grischard** [comment]
> Although we could mitigate this with the secure attribute on cookies

If all map pages aren't redirected to https, this would be confusing. "Why can't I edit? I've just logged in!"

**Comment by zerebubuth** [comment]
> If all map pages aren't redirected to https, this would be confusing. "Why can't I edit? I've just logged in!"

Logging in switches the scheme to HTTPS, so this wouldn't be a natural navigation flow for the user. Clicking the edit button could also redirect to HTTPS, as one needs to be logged in for that anyway.

**Comment by grischard** [comment]
> HSTS does not mean that it is impossible to use the site with HTTP. It
only means that once someone accessed the site via HTTPS they will
continue to use HTTPS for a certain amount of time.

Section 7.2. of the HSTS RFC says: "An HSTS Host MUST NOT include the STS
header field in HTTP responses conveyed over non-secure transport."

A while ago, I asked Jeff Hodges, one of the RFC's author, why that was. If you provide both HTTP and HTTPS for compatibility reasons, HSTS on HTTP responses would seem to be a neat way of encouraging HTTPS without forcing it.

Jeff replied:

> The rationale for not including the STS header field in unsecured HTTP responses is that it (the sts header field) is a security directive and http clients must not heed such directives obtained over unsecured channels. So it is best if webapps do not issue the STS header over unsecured channels. We made it a "MUST NOT" to try to get folks' attention. it is much more critical for the clients to get this right, as stipulated in section 8.1.

**Comment by grischard** [comment]
Hitting the 'Edit' button to load the map in JOSM from an https page can apparently sometimes fail to work. (Can anyone reproduce this?)

I remember running into that and fixing it, although I don't remember how. Osmose had the same problem and [uses a redirect hack](https://github.com/osm-fr/osmose-frontend/issues/36).

**Comment by SomeoneElseOSM** [comment]
JOSM HTTPS remote control support depends on Java trusting a self-signed certificate for "localhost" doesn't it?  Presumably there are going to be a bunch of people out there who don't have write access to the Java keystore (or who just are going to be put off by the scary message that you get when enabling remote control HTTPS support).

Obviously JOSM remote control isn't the only way to get to start editing in an area (FWIW I tend to start via "download object" or locating somewhere on the imagery).

**Comment by tyll** [comment]
On Tue, Nov 08, 2016 at 08:51:25AM -0800, Guillaume Rischard wrote:

> A while ago, I asked Jeff Hodges, one of the RFC's author, why that
> was. If you provide both HTTP and HTTPS for compatibility reasons,
> HSTS on HTTP responses would seem to be a neat way of encouraging
> HTTPS without forcing it.

There is a different technique to notify that HTTPS is preferred:

http://stackoverflow.com/questions/31950470/what-is-the-upgrade-insecure-requests-http-header/32003517#32003517

Also including an (hidden) image via HTTPS and sending an HSTS header
there can be used to upgrade clients to HTTPS if they support HTTPS and
HSTS.

**Comment by pnorman** [comment]
> There is a different technique to notify that HTTPS is preferred: ...

I think the techniques should wait until after the policy discussion.

**Comment by Komzpa** [comment]
HTTP 2.0 for tiles can move them away from [abcd].tile.osm.org scheme, keeping less connections, allowing browser to request all the tiles at once and get them in single round-trip. From my experience with Wargaming GlobalMap, this is especially good thing for slow / laggy connections. This requires HTTPS.

Ensuring HTTPS via HSTS for tiles will likely make sure browsers will send Referer header to the server (they skip it when loading http image into https page), which allows more precise traffic source tracking.

**Comment by Klumbumbus** [comment]
>Hitting the 'Edit' button to load the map in JOSM from an https page can apparently sometimes fail to work. (Can anyone reproduce this?)

In Firefox remote control to JOSM via https is not working at all. Only http works. More info at
- https://github.com/hotosm/osm-tasking-manager2/issues/735
- https://josm.openstreetmap.de/ticket/10033
- https://josm.openstreetmap.de/ticket/10230

**Comment by SomeoneElseOSM** [comment]
Not something that affects the main map site, but I was trying to translate https://lists.openstreetmap.org/pipermail/talk-es/2016-November/014568.html from Spanish to English for the OSM weekly news.  Google Translate screws up the formatting somewhat, so I thought I'd try http://www.bing.com/translator - and that says "Translation of secure pages is not supported".  This is a problem we have now rather than a theoretical future one as OSM's pipermail site redirects from http to https already.

**Comment by scaidermern** [comment]
> In Firefox remote control to JOSM via https is not working at all. Only http works.

Works fine here. I always use HTTPS with Firefox on openstreetmap.org (thanks to HTTPS Everywhere). To get JOSM's remote control to work via HTTPS you have to open https://127.0.0.1:8112 in your browser and permanently store an exception for this self-signed certificate. Afterwards the remote control should work.

**Comment by Klumbumbus** [comment]
OK, thanks. Works for me.

**Comment by grischard** [comment]
Relevant for discussion: [The Guardian has moved to HTTPS](https://www.theguardian.com/info/developer-blog/2016/nov/29/the-guardian-has-moved-to-https) today.

**Comment by simonpoole** [comment]
@grischard the guardian just happens to be the newspaper site that requires you to leak what you are reading to something roughly between 15 and 30 third party sites, always including at least facebook but most of the other usual suspects too. It is hard to imagine a site that has -less- relevant privacy policies than them.

**Comment by grischard** [comment]
[Extensive write-up on moving to HTTPS only from the Stack Overflow operations team](https://nickcraver.com/blog/2017/05/22/https-on-stack-overflow/).

**Comment by grischard** [comment]
So where are we six months later?

More and more sites are enabling HTTPS by default, so HTTPS should be blocked in less places, and be more reliable. There will also be even less legacy browsers.

JOSM still doesn't reliably add its certificate to the certificate stores on all browsers and all platforms. The https remote isn't enabled by default in JOSM. These three tickets must be resolved before we can switch:

1. [Enable https support on remote by default](https://josm.openstreetmap.de/ticket/14398)
2. [Mention that self-signed certificate might have to be accepted for remote control to work
](https://josm.openstreetmap.de/ticket/12230)
3. [Josm do not start in remote control from osm.org in https
](https://josm.openstreetmap.de/ticket/10033), which is about adding JOSM's certificate to the keystore/keychain/NSS.

**Comment by grischard** [comment]
Simon points out https://github.com/MarcusWolschon/osmeditor4android/issues/523 - Android versions up to at least 6.0 have issues with SNI.

**Comment by Bisaloo** [comment]
Just a heads up: from now on (Firefox 55.0), Firefox requires HTTPS for the geolocation API as well.

[Firefox 55.0 release notes](https://www.mozilla.org/en-US/firefox/55.0/releasenotes/)

[Can I use... Geolocation](https://caniuse.com/#search=geolocation)

**Comment by systemed** [comment]
On the Remote Control side, it also looks possible that desktop Potlatch will support Remote Control over HTTP but not HTTPS, for Adobe AIR reasons. I guess there are potential ways around this, such as an openstreetmap-website user setting to "Launch Remote Control over unencrypted local connection" which would use a popup or redirect rather than the current appended iframe.

**Comment by rugk** [comment]
[Does my site need HTTPS?](https://doesmysiteneedhttps.com/)

[Still think you don't need HTTPS?](https://scotthelme.co.uk/still-think-you-dont-need-https/)

[Google: Why you should always use HTTPS](https://developers.google.com/web/fundamentals/security/encrypt-in-transit/why-https)

[Is TLS fast yet?](https://istlsfastyet.com/)

Breaking OAuth for HTTPS is also a good thing, as any authentication is worth nothing when it is done using insecure protocols, such as HTTP.

**Comment by Bisaloo** [comment]
Blocking: https://github.com/hotosm/osm-tasking-manager2/issues/988

**Comment by pnorman** [comment]
> Blocking: hotosm/osm-tasking-manager2#988

I don't see that blocks the OSMF servers from forcing HTTPS.

**Comment by Bisaloo** [comment]
Maybe blocking is a strong word but `https` on openstreetmap.org breaks stuff from hotosm (and similar sites) due to MCB.

Forcing `https` will make this problem worse. *Unless* `task.hotosm.org` and others get `https` support first.

**Comment by grischard** [comment]
So the issue in general is that content that the iD editor embeds from other sites will have to be served over https. Solving hotosm/osm-tasking-manager2#988 and sending https gpx links is the way to solve that.

Https usage on openstreetmap.org isn't uncommon, and showing the gpx boundary from hot is already broken for all those users. I'm actually surprised that tasks.hotosm.org isn't available over https at all.   

I've added this issue to the list.

**Comment by SomeoneElseOSM** [comment]
@grischard Apologies if I'm missing something here, but how is the fact that tasks.hotosm.org doesn't support https at all relevant to whether it's a good idea to force users of openstreetmap.org to use https whether they want to or not?

**Comment by grischard** [comment]
@SomeoneElseOSM openstreetmap.org doesn't exist in a vacuum; it'd be relevant if forcing https broke a lot of things for the mappers, just like hsts breaking oauth is. The fact that it took almost one year for this to come up makes me think that it won't, and the issue can easily be mitigated on hot's side anyway, but I want to be thorough in looking at the possible issues.

**Comment by rugk** [comment]
Maybe depreciate/announce that you are removing HTTP support and let sites some time to switch to HTTPS. Just note that OAUTH and any password/authentication over insecure channels (HTTP) is not secure. Everyone can sniff your password, so…
In any case the aim should at some point be HTTPS-only. Before that you may redirect to HTTPS e.g. only on the main site or on most sites (e.g. login page) with some exceptions.

**Comment by HolgerJeromin** [comment]
> Just note that OAUTH and any password/authentication over insecure channels (HTTP) is not secure. Everyone can sniff your password, so…

With oauth no passwort is transmitted

**Comment by rugk** [comment]
But session tokens, or some other identifier. Any MITM possibility is bad. And you still submit passwords over http, so we don't even need to argue about oauth.

**Comment by simonpoole** [comment]
@rugk please get your facts straight, if you are using OAuth as essentially all major editors do, you do NOT authorize (aka login once with login/password) over http, but over https.

**Comment by bhousel** [comment]
@simonpoole - @rugk is correct that OAuth secrets can be transmitted over http when embedding a browser editor like Potlatch or iD on the openstreetmap website, if the website was first loaded via http.  There is no really secure way to do communication between the embedding site and the editor iframe.

**Comment by simonpoole** [comment]
@bhousel nobody way claiming anything else, the authorisation process however takes place over https, just as I wrote.

**Comment by don-vip** [comment]
Hello,
From https://josm.openstreetmap.de/ticket/10033#comment:38 I discover that all recent browsers now accept `http://127.0.0.1:8111` from https (or will very soon).
So JOSM is no longer a reason to block https by default on osm.org. I even plan to remove https support on port 8112 as it does not work properly everywhere and appears to be useless now.

**Comment by grischard** [comment]
What do you all think of already adding HSTS to [the tile server](https://github.com/openstreetmap/chef/blob/master/cookbooks/tile/templates/default/apache.erb)? 

AFAICT this wouldn't break anything, and modern browsers already upgrade to http2 with tls anyway. We would also hear back from hypothetical users for whom https is completely blocked.

**Comment by grischard** [comment]
For those of us only following this ticket, see #190 also.

**Comment by nmxcgeo** [comment]
Hi, 

I saw today, that you still send out e-mails with http-links for e.g. comments on your changesets / notes / whatever. As yo are already serving HSTS headers for www.openstreetmap.org I think it's reasonable to change these links to https as well.

Greetings
Nmxcgeo

**Comment by grischard** [comment]
Maybe we could revisit https://github.com/openstreetmap/openstreetmap-website/pull/1341 and https://github.com/openstreetmap/openstreetmap-website/pull/939. What do you think @tomhughes ?

**Comment by tomhughes** [comment]
I'm not sure we can merge either of those as is, as lots of other people use our code base and they may not be running on an https capable site so we probably need to make it configurable in some way.

I'll have a think about it...

**Comment by HolgerJeromin** [comment]
At least https://github.com/openstreetmap/openstreetmap-website/pull/939 can be done as [Protocol-relative URL](https://www.paulirish.com/2010/the-protocol-relative-url/)

**Comment by tomhughes** [comment]
Not really - that is code designed to be cut and pasted on a third party site and that would go wrong if it was pasted on an https site and referenced an http only site.

**Comment by HolgerJeromin** [comment]
Sorry, you are right. Thanks for the explanation.

**Comment by pnorman** [comment]
We're using https by default now.

**Comment by genodeftest** [comment]
Thank you very much!
PS: Your implementation looks pretty nice! See these good ratings
* SSLlabs: A+ https://www.ssllabs.com/ssltest/analyze.html?d=openstreetmap.org&s=130.117.76.7&latest
* Securityheaders: https://securityheaders.com/?q=openstreetmap.org&followRedirects=on

**Comment by grischard** [comment]
Please re-open, tile is indeed the last one that still doesn't redirect.

How to reproduce:

`curl -v a.tile.openstreetmap.org/10/11/12.png`

Expected:

redirect

Actual result:

200 OK and PNG

**Comment by rugk** [comment]
> support modern […] http/2

I think this should be enough reason to upgrade it. I mean, [especially for multiple connections/tiles like this](https://www.httpvshttps.com/) it should improve speed very much…

**Comment by grischard** [comment]
Yeah, we have an nginx reverse proxy in front of squid, but only for https.

**Comment by pnorman** [comment]
Now that we've moved the tile CDN, I think it would be easy to redirect HTTP traffic to HTTPS. This would not impact any users of openstreetmap.org because they would have already browsed to tile.openstreetmap.org and gotten HSTS headers. What it would impact is apps and browsers using a site that has HTTP tiles where the browsers haven't seen HTTPS tiles.

**Comment by grischard** [comment]
It's even better than that. With HSTS preloading, the only remaining http users are apps, bots, antique browsers more than six years old, and misconfigured reverse proxies.

We could detect user agents that claim to be modern browsers, and block them, since there is no way they would not use https.

**Comment by Firefishy** [comment]
`tile.openstreetmap.org`, `a.tile.openstreetmap.org`, `b.tile.openstreetmap.org`, `c.tile.openstreetmap.org` are on the HSTS preloading published list. https://hstspreload.org/?domain=tile.openstreetmap.org

**Comment by Firefishy** [comment]
We keep HTTP enabled for backward compatibility and to catch apps which are faking web browsers. A HTTP -> HTTPS redirect would be insecure to interception and doubles the number of requests we'd have to process. HSTS preloading is a better route to achieve the same security.

**Comment by Firefishy** [comment]
Above I conflated this ticket being for HTTPS by default on tile.openstreetmap.org instead of openstreetmap.org. www.openstreetmap.org is already HTTPS default.

**Comment by rugk** [comment]
Okay, awesome! I agree with all you've said in your last comments, so I think, yes, this issue is solved and I'd thus suggest to close it.

**Comment by genodeftest** [comment]
Thank you very much!
